---
title: "Introduction"
output: html_document
vignette: >
  %\VignetteIndexEntry{Introduction}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=7,
                      echo=TRUE, warning=FALSE, message=FALSE,
                      tidy=TRUE, collapse = TRUE, comment = "#>")
```

```{r setup}
set.seed(1234)
library(panstripe)
library(ggplot2)
```

#Simulate a pangenome

We first simulate a pangenome presence/absence matrix and corresponding core genome phylogeny.

```{r}
sim <- simulate_pan(rate = 1e-3)
```

#Fit gene gain/loss model

We can now estimate the expected number of gene gain and loss events along each branch of the phylogeny and fit a tweedie regression model.

```{r}
fit <- panstripe(sim$pa, sim$tree, quiet=TRUE)
```

#Investigate results

We can test whether the expected gene gain/loss rate is significantly different from a model consisting of just errors by looking at the p-value of the fit.

```{r}
fit$summary
```


We can also generate a plot of the resulting fit with confidence intervals generated via bootstrap sampling of the phylogeny branches.

```{r}
plot_pangenome_fits(fit)
```

#Comparing pangenomes

For simplicity we consider a comparison with a simulated closed pangenome. This could reflect a species with a very stable set of genes such as *M. tuberculosis*.

```{r}
sim_closed <- simulate_pan(rate = 0)
```

Again we fit the a tweedie regression model after first estimating the gene gain/loss events per branch.

```{r}
fit_closed <- panstripe(sim_closed$pa, sim_closed$tree, quiet=TRUE)
fit_closed$summary
```

We can now test to see whether the choice of pangenome (open or closed) is significant.

```{r}
compare_pangenomes(fit, fit_closed)
```

We can also plot the difference between the fits.

```{r}
plot_pangenome_fits(list(
  open=fit, 
  closed=fit_closed), boot = TRUE)
```

#Investigate differences in the average size of gene gain/loss events

Sometimes we would like to know if it is the size rather than the rate of recombination events that is driving an apparent different in the rate of gene gain/loss. As we are fitting a tweedie regression model it is possible to tease apart the driving force of apparent accessory rate differences.


###Simulate

First lets simulate and fit two pangenomes. One with a larger average recombination size (in number of genes).

```{r}
sim_small_rec <- simulate_pan(rate = 1e-3, mean_trans_size = 2)
sim_large_rec <- simulate_pan(rate = 1e-3, mean_trans_size = 10)

res_small_rec <- panstripe(sim_small_rec$pa, sim_small_rec$tree, quiet=TRUE)
res_large_rec <- panstripe(sim_large_rec$pa, sim_large_rec$tree, quiet=TRUE)
```

We can first test whether the overall rate of gene gain/loss is different between the two pangenomes.

```{r}
comp <- compare_pangenomes(res_small_rec, res_large_rec)
comp$summary
```

As expected given the large difference in the average size of gene gain/loss events there is a significant difference in the total rate between the two pangenomes.

We can now look closer into what is driving this difference.

```{r}
plot_pangenome_fits(list(
  large=res_large_rec, 
  small=res_small_rec))
```

We can look at how the estimated parameters for the Gamma and Poisson distributions vary with the core branch distance.

```{r}
plot_dist_params(list(
  large=res_large_rec, 
  small=res_small_rec))
```


## References

---
nocite: '@*'
...

##Appendix

```{r}
sessionInfo()
```
